src/utils/ParseRouteUtilsV3.js

```diff
@@ -0,0 +1,680 @@
+ * - Checks user can make the bus
+ * - Checks directions not excessively long
+ * @param route
+ * @param startCoords
+ * @param endCoords
+ * @param maxWalkingDistance
+ * @param departureDelayBuffer
+ * @param departureTimeNowMs
+ * @returns {Object}
+ */
+async function condenseRoute(
+  route: Object,
+  startCoords: Object,
+  endCoords: Object,
+  maxWalkingDistance: number,
+  departureDelayBuffer: boolean,
+  departureTimeNowMs: number,
+) {
+  await trimFirstLastDirections(route, startCoords, endCoords);
+
+  let previousDirection = null;
+  let totalDistanceWalking = 0;
+
+  for (let index = 0; index < route.directions.length; index++) {
+    const direction = route.directions[index];
+    const startTime = Date.parse(direction.startTime);
+    const endTime = Date.parse(direction.endTime);
+
+    // Discard routes with directions that take over 2 hours time
+    if (startTime + (ONE_HOUR_IN_MS * 2) <= endTime) {
+      return null;
+    }
+
+    /**
+     * Discard routes where not possible to walk to bus given departure buffer
+     * We subtract ONE_MIN_IN_MS from departureTimeNow to avoid the
+     * case where we don't show a bus leaving at the same minute as departureTime.
+     * We also have to account for the delay time because if a bus is delayed, we want to display
+     * the route even if the start time is earlier than the current time.
+     */
+    if (departureDelayBuffer && direction.type === DIRECTION_TYPE.DEPART) {
+      // direction.delay could be undefined, NaN, or null
+      const delayMs = direction.delay ? direction.delay * 1000 : 0;
+      if (startTime < departureTimeNowMs - ONE_MIN_IN_MS - delayMs) {
+        return null;
+      }
+    }
+
+    if (previousDirection
+      && previousDirection.type === DIRECTION_TYPE.DEPART
+      && direction.type === DIRECTION_TYPE.DEPART) {
+      /*
+       * Discard route if a depart direction's last stopID is not equal to the next direction's first stopID,
+       * Fixes bug where graphhopper directions are to get off at a stop and get on another stop
+       * far away with no walking direction in between, EG: 1. get off at Statler 2. board at RPCC
+       */
+      if (previousDirection.stops[previousDirection.stops.length - 1].stopID
+        !== direction.stops[0].stopID) {
+        return null;
+      }
+
+      /*
+       * If consecutive bus directions have the same routeNumber,
+       * then replace the last direction with merged directions and remove the current direction.
+       * No real transfer, probably just change in trip_ids.
+       */
+      if (previousDirection.routeNumber === direction.routeNumber) {
+        route.directions.splice(index - 1, 2, mergeDirections(previousDirection, direction));
+        index -= 1; // since we removed an element from the array
+      }
+
+      // Discard routes with over 1 hours time waiting between each direction
+      if (previousDirection) {
```

Comment:
isn't this in the big nested if that checked previousDirection before entering? why are you doing another check